### PR TITLE
fix: sanitize changes diagram input

### PR DESCRIPTION
**Pull request by nagai9999**
### **User description**
Issue: https://github.com/qodo-ai/pr-agent/issues/2211

### **PR Type**
Enhancement, Tests

___

### **Description**
- Extract diagram sanitization logic into dedicated function

- Remove backticks from node labels in mermaid diagrams

- Add comprehensive unit tests for diagram sanitization

- Handle missing closing fence and invalid diagram formats

___

### Diagram Walkthrough

```mermaid
flowchart LR
  input["Raw diagram input"] -- "sanitize_diagram()" --> validation["Validate fence markers"]
  validation -- "Fix missing closing" --> cleanup["Remove backticks from labels"]
  cleanup -- "Return sanitized" --> output["Sanitized diagram string"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Extract and enhance diagram sanitization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_description.py

<ul><li>Refactored diagram sanitization logic into new <code>sanitize_diagram()</code> <br>function<br> <li> Function validates fence markers, fixes missing closing backticks, and <br>removes backticks from node labels<br> <li> Updated <code>_prepare_data()</code> to use new sanitization function instead of <br>inline logic<br> <li> Returns empty string for invalid diagrams that don't start with fence <br>markers</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2212/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+19/-5</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_pr_description.py</strong><dd><code>Add unit tests for diagram sanitization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unittest/test_pr_description.py

<ul><li>Added new test file with comprehensive unit tests for diagram <br>sanitization<br> <li> Tests cover invalid diagrams without fence markers, missing closing <br>fences, and backtick removal<br> <li> Tests verify backticks are only removed from node labels, not from <br>edge labels<br> <li> Tests confirm normal diagrams are processed correctly with newline <br>prefix</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2212/files#diff-ad3dc37aacedd574aa6da85944e3afcec9316d215f8b12b966289de4a477bb6d">+69/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Extract diagram sanitization logic into dedicated `sanitize_diagram()` function

- Fix missing closing fence markers and remove backticks from node labels

- Add comprehensive unit tests covering edge cases and validation

- Validate diagram input format and handle invalid diagrams gracefully

___

### Diagram Walkthrough

```mermaid
flowchart LR
  raw["Raw diagram input"]
  validate["Validate fence markers"]
  fix["Fix missing closing fence"]
  clean["Remove backticks from labels"]
  output["Sanitized diagram string"]
  raw -- "sanitize_diagram()" --> validate
  validate --> fix
  fix --> clean
  clean --> output
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Extract and enhance diagram sanitization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_description.py

<ul><li>Extracted diagram sanitization logic into new <code>sanitize_diagram()</code> <br>function<br> <li> Function validates fence markers, fixes missing closing backticks, and <br>removes backticks from node labels using regex<br> <li> Updated <code>_prepare_data()</code> to call new sanitization function instead of <br>inline logic<br> <li> Returns empty string for invalid diagrams that don't start with fence <br>markers</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2212/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+21/-5</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_pr_description.py</strong><dd><code>Add unit tests for diagram sanitization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unittest/test_pr_description.py

<ul><li>Added new test file with comprehensive unit tests for <br><code>sanitize_diagram()</code> function<br> <li> Tests cover invalid diagrams without fence markers, missing closing <br>fences, and backtick removal from labels<br> <li> Tests verify backticks are only removed from node labels, not from <br>edge labels<br> <li> Tests confirm normal diagrams are processed correctly with newline <br>prefix and handle non-string inputs</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2212/files#diff-ad3dc37aacedd574aa6da85944e3afcec9316d215f8b12b966289de4a477bb6d">+75/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Review by naorpeled — APPROVED** [review]

File: pr_agent/tools/pr_description.py [hunk 1]
```diff
@@ -771,6 +769,24 @@ def add_file_data(self, delta_nbsp, diff_plus_minus, file_change_description_br,
...
+    if not diagram.startswith('```'):
+        return ''
+    
+    # fallback missing closing
+    if not diagram.endswith('```'):
+        diagram += '\n```'  
+    
+    # remove backticks inside node labels: ["`label`"] -> ["label"]
+    lines = diagram.split('\n')
+    result = [re.sub(r'\["([^"]*?)"\]', lambda m: '["' + m.group(1).replace('`', '') + '"]', line) for line in lines]
```
[the remark this answers is not in the document]
Comment (nagai9999) [reply pr_agent/tools/pr_description.py]: Fixed in be9fe13 and 9251f62.

- Replaced the one-liner list comprehension with a `for` loop + multi-line `re.sub` call to stay within the 120-char limit
- Also tightened the fence check from `` startswith('```') `` to `` startswith('```mermaid') ``

**Comment by nagai9999** [comment]
Hi @naorpeled, all Qodo comments have been addressed.
